tests/python/pants_test/util/test_meta.py

```diff
@@ -31,9 +31,101 @@ def method(self):
       AbstractMethod()
 
 
-class SingletonTest(unittest.TestCase):
+class SingletonTest(TestBase):
   def test_singleton(self):
     class One(Singleton):
       pass
 
     self.assertIs(One(), One())
+
+
+class WithProp(object):
+  _value = 3
+
+  @classproperty
+  def some_property(cls):
+    "some docs"
+    return cls._value
+
+
+class OverridingValueField(WithProp):
+  _value = 4
+
+
+class OverridingMethodDefSuper(WithProp):
+
+  _other_value = 2
+
+  @classproperty
+  def some_property(cls):
+    return super(OverridingMethodDefSuper, cls).some_property + cls._other_value
+
+
+class ClassPropertyTest(TestBase):
+  # TODO: The assertions on both the class and an instance of it might not be necessary, if class
+  # attributes are assumed to always be the same on their instances.
```

Comment:
I think it's entirely possible (and an existing pattern in pants' code base) that an instance could inadvertently override a class variable - so having a test for that case up front would make a lot of sense to me - if for nothing else to clarify the expected behavior here.

i.e. I would expect this test to pass:

```
[omerta pants2 (pr5901)]$ git diff | cat
diff --git a/tests/python/pants_test/util/test_meta.py b/tests/python/pants_test/util/test_meta.py
index 4de0da6..d91e12a 100644
--- a/tests/python/pants_test/util/test_meta.py
+++ b/tests/python/pants_test/util/test_meta.py
@@ -53,6 +53,11 @@ class OverridingValueField(WithProp):
   _value = 4
 
 
+class OverridingValueInit(WithProp):
+  def __init__(self, v):
+    self._value = v
+
+
 class OverridingMethodDefSuper(WithProp):
 
   _other_value = 2
@@ -77,6 +82,9 @@ class ClassPropertyTest(TestBase):
     self.assertEqual(4, OverridingValueField.some_property)
     self.assertEqual(4, OverridingValueField().some_property)
 
+  def test_override_inst_value(self):
+    self.assertEqual(4, OverridingValueInit(3).some_property)
+
   def test_override_method_super(self):
     self.assertEqual(5, OverridingMethodDefSuper.some_property)
     self.assertEqual(5, OverridingMethodDefSuper().some_property)
```

but noting that it does not currently appear to work as expected:

```
 ==================== FAILURES ====================
 ___ ClassPropertyTest.test_override_inst_value ___
 
 self = <pants_test.util.test_meta.ClassPropertyTest testMethod=test_override_inst_value>
 
     def test_override_inst_value(self):
 >     self.assertEqual(4, OverridingValueInit(3).some_property)
 E     AssertionError: 4 != 3
 
 .pants.d/pyprep/sources/2a69993898e39901cfba78c09ed3eaea1cc1406f/pants_test/util/test_meta.py:86: AssertionError
  generated xml file: /Users/kwilson/dev/pants2/.pants.d/test/pytest/tests.python.pants_test.util.meta/junitxml/TEST-tests.python.pants_test.util.meta.xml 
 
 
 ====== 1 failed, 11 passed in 0.86 seconds =======
```

either way, cementing this in with a test (whether xfail or not) and some clarifying docs/notes about the behavior in this situation would probably be good.

Reply:
Hm. I think it would make sense for that test to pass, but I get the same behavior you describe when using an `@classmethod`. It feels like doing whatever `@classmethod` does would be the least surprising thing for `@classproperty` to do -- even if it's potentially surprising to silently switch to using the instance field. Does that seem reasonable?

I'm adding your test case, as well as a couple assertions for an `@classmethod` to confirm that the behavior is the same when accessed on an instance. I am also going to add an explanation of this behavior in the `@classproperty` doc string.

Reply:
Let me know if 31e1a38 is the kind of documentation/testing you were thinking of.

Reply:
> accessed through an instance and not through the class

Phrasing it like this made it more clear to me. I made a hacky subclass of the pex `Installer` at one point which overrode an `@classmethod` with an instance method -- this ended up getting deleted in favor of a different approach but even though shadowing *could* happen by accident, that is also a feature that can be used intentionally (I recognize that overriding the method is a little different than making a shadowed field).

Reply:
works for me!

(ultimately, I don't have a strong view either way - the concern here is just that we have a documented stance on that behavior vs someone with mismatched expectations stumbling upon a TODO)

Reply:
Yes! I really appreciate you pointing this out @kwlzn because I can see a future self getting highly flummoxed without taking the time to make it clear here. I've removed that TODO in 81d3d6c -- let me know if the `@classproperty` docstring added in 31e1a38 is sufficient documentation, or if there could be a note elsewhere.